File: tower_of_hanoi_optimised.cpp
#include <iostream>

void moveRing(long rings[], long& moves, long ringIndex, long target);
void moveRing(long rings[], long ringIndex, long target);

int main(int argc, char const *argv[]) {
    using namespace std;
    int heightInput;
    // cout << "Please enter the number of rings in your Tower of Hanoi puzzle." << endl;
    // cin >> heightInput;
   if (argc > 1) {
     heightInput = atoi(argv[1]);
   } else {
     heightInput = 12;
   }

    // cout << heightInput << endl;
    long rings[heightInput + 1];
    for (int i = 0; i < heightInput; i++) {
        rings[i] = 1;
    }
    rings[heightInput] = 0;
    // for (int i = 0; i < heightInput; i++) {
    //     cout << rings[i] << " ";
    // }
    // cout << endl;
    if (argc > 2) {
      long moves = 0;
      moveRing(rings, moves, 0, 3);
      cout << "Moves: " << moves << endl;
    } else {
      moveRing(rings, 0, 3);
    }

    // for (int i = 0; i < heightInput; i++) {
    //     cout << rings[i] << " ";
    // }
    // cout << endl;

    return 0;
}

void moveRing(long rings[], long& moves, long ringIndex, long target){
  if (rings[ringIndex + 1])
      moveRing(rings, moves, (ringIndex + 1), 6 - target - rings[ringIndex]);
  rings[ringIndex] = target;
  // cout << "Moved ring " << ringIndex << " from " << home << " to " << target << "." << endl;
  moves++;
  if (rings[ringIndex + 1])
      moveRing(rings, moves, (ringIndex + 1), target);
}

void moveRing(long rings[], long ringIndex, long target){
  if (rings[ringIndex + 1])
      moveRing(rings, ringIndex + 1, 6 - target - rings[ringIndex]);
  rings[ringIndex] = target;
  // cout << "Moved ring " << ringIndex << " from " << home << " to " << target << "." << endl;
  if (rings[ringIndex + 1])
      moveRing(rings, ringIndex + 1, target);
}

File: tower_of_hanoi_optimised.js
"use strict";

const towerOfHanoi = function (height) {
  const moveRing = function (ring, target) {
    // const home = rings[ring];
    // const spare = [1, 2, 3].find((tower) => {
    //   return tower !== target && tower !== home;
    // });
    if (rings[ring + 1]) {
          moveRing((ring + 1), 6 - target - rings[ring]);
    }
    rings[ring] = target;
    // console.log(`Moved ring ${ring} from ${home} to ${target}.`);
    moves++;
    if (rings[ring + 1]) {
      moveRing((ring + 1), target);
    }
  }

  let moves = 0;
  const rings = [];
  for (let i = 0; i < height; i++) {
    rings[i] = 1;
  };
  // console.log(rings);
  moveRing(0, 3);
  // console.log(rings);
  console.log(`Moves: ${moves}`);
}
if (process.argv.length > 2) {
  towerOfHanoi(process.argv[2]);
} else {
  towerOfHanoi(12);
}
